yeat/cli/config.py

```diff
@@ -8,44 +8,40 @@
 # -------------------------------------------------------------------------------------------------
 
 import json
+from pathlib import Path
 from warnings import warn
 
 
-ALGORITHMS = ("spades", "megahit", "unicycler", "flye", "canu")
-KEYS = ["algorithm", "extra_args"]
+ALGORITHMS = ["spades", "megahit", "unicycler", "flye", "canu"]
+KEYS1 = ["samples", "assemblers"]
+KEYS2 = ["label", "algorithm", "extra_args", "samples"]
 
 
 class AssemblyConfigurationError(ValueError):
     pass
 
 
 class AssemblerConfig:
-    def __init__(self, algorithm, extra_args=""):
-        if algorithm not in ALGORITHMS:
-            raise AssemblyConfigurationError(f"Unsupported assembly algorithm '{algorithm}'")
-        self.algorithm = algorithm
-        self.extra_args = extra_args
-
-    @classmethod
-    def from_json(cls, data):
-        return cls(data["algorithm"], data["extra_args"])
-
     @staticmethod
     def parse_json(instream):
         configdata = json.load(instream)
-        for entry in configdata:
-            AssemblerConfig.validate(entry)
-        algorithms = [entry["algorithm"] for entry in configdata]
-        if len(algorithms) > len(set(algorithms)):
-            message = "Duplicate assembly configuration: please check config file"
+        AssemblerConfig.validate(configdata, KEYS1)
+        for assembler in configdata["assemblers"]:
+            AssemblerConfig.validate(assembler, KEYS2)
+        labels = [assembler["label"] for assembler in configdata["assemblers"]]
+        if len(labels) > len(set(labels)):
+            message = "Duplicate assembly labels: please check config file"
             raise AssemblyConfigurationError(message)
-        configlist = [AssemblerConfig.from_json(entry) for entry in configdata]
-        return configlist
+        sampledict = {}
+        for key, value in configdata["samples"].items():
+            sampledict[key] = Sample.from_json(value)
+        assemblerlist = [Assembler.from_json(assembler) for assembler in configdata["assemblers"]]
+        return sampledict, assemblerlist
 
     @staticmethod
-    def validate(config):
-        missingkeys = set(KEYS) - set(config.keys())
-        extrakeys = set(config.keys()) - set(KEYS)
+    def validate(config, keystemp):
+        missingkeys = set(keystemp) - set(config.keys())
+        extrakeys = set(config.keys()) - set(keystemp)
```

Comment:
It makes sense to reuse this code for both the top-level configurations as well as the lower-level assembler configurations. However, it's poor design to outsource the coordination of the validation. Whatever code wants to validate the config has to validate the top-level config and then iterate over all the subconfigs and validate them separately. That code has to import the `KEYS1` and `KEYS2` constants as pass them as arguments. Why not hide all those details internally and give the user a simple interface?

Reply:
Updated the `validate()`. The function is called within the `AssemblerConfig` class `__init__()`.